File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/page/storage/PageLoader.kt
package com.crescentflare.piratesgame.page.storage

import android.content.Context
import kotlinx.coroutines.*
import okhttp3.*
import java.io.IOException
import java.lang.ref.WeakReference

/**
 * Page storage: handles loading of pages from several sources
 */
class PageLoader(context: Context, location: String) {

    // ---
    // Members
    // ---

    private val context: Context
    private val location: String
    private val loadInternal: Boolean
    private var continuousCompletion : WeakReference<PageLoaderContinuousCompletion>? = null
    private var loading = false
    private var waiting = false


    // ---
    // Initialization
    // ---

    init {
        this.context = context
        this.location = location
        loadInternal = !location.contains("://")
    }


    // ---
    // Loading
    // ---

    fun load(completion: (page: Page?, exception: Throwable?) -> Unit) {
        val cachedPage = PageCache.instance.getEntry(location)
        if (cachedPage != null) {
            completion(cachedPage, null)
        } else if (loadInternal) {
            loadInternal {
                if (it != null) {
                    PageCache.instance.storeEntry(location, it)
                }
                completion(it, null)
            }
        } else {
            loadOnline("ignore") { page, exception ->
                if (page != null) {
                    PageCache.instance.storeEntry(location, page)
                }
                completion(page, exception)
            }
        }
    }

    private fun loadOnline(currentHash: String, completion: (page: Page?, exception: Throwable?) -> Unit) {
        if (!loadInternal) {
            val client = OkHttpClient()
            loading = true
            client.newCall(Request.Builder().url(location).header("X-Mock-Wait-Change-Hash", currentHash).build()).enqueue(object : Callback {
                override fun onFailure(call: Call, exception: IOException) {
                    GlobalScope.launch(Dispatchers.Main, CoroutineStart.DEFAULT) {
                        loading = false
                        completion(null, exception)
                    }
                }

                @Throws(IOException::class)
                override fun onResponse(call: Call, response: Response) {
                    val jsonString = response.body()?.string()
                    if (jsonString != null) {
                        val page = Page(jsonString)
                        GlobalScope.launch(Dispatchers.Main, CoroutineStart.DEFAULT) {
                            loading = false
                            completion(page, null)
                        }
                    } else {
                        GlobalScope.launch(Dispatchers.Main, CoroutineStart.DEFAULT) {
                            loading = false
                            completion(null, null)
                        }
                    }
                }
            })
        } else {
            completion(null, null)
        }
    }

    private fun loadInternal(completion: (page: Page?) -> Unit) {
        loading = true
        GlobalScope.launch(Dispatchers.Default, CoroutineStart.DEFAULT) {
            val page = loadInternalSync()
            GlobalScope.launch(Dispatchers.Main, CoroutineStart.DEFAULT) {
                loading = false
                completion(page)
            }
        }
    }

    fun loadInternalSync(): Page? {
        if (loadInternal) {
            try {
                val stream = context.getAssets().open("pages/${location}")
                val jsonString = stream.bufferedReader().use { it.readText() }
                return Page(jsonString)
            } catch (ignored: IOException) {
            }
        }
        return null
    }


    // ---
    // Continuous loading
    // ---

    fun startLoadingContinuously(completion: PageLoaderContinuousCompletion) {
        continuousCompletion = WeakReference(completion)
        tryNextContinuousLoad()
    }

    fun stopLoadingContinuously() {
        continuousCompletion = null
    }

    private fun tryNextContinuousLoad() {
        if (!loading && !waiting && continuousCompletion != null) {
            if (loadInternal) {
                loadInternal {
                    if (it != null) {
                        PageCache.instance.storeEntry(location, it)
                        continuousCompletion?.get()?.didUpdatePage(it)
                    }
                }
            } else {
                val hash = PageCache.instance.getEntry(location)?.hash ?: "unknown"
                loadOnline(hash) { page, _ ->
                    var waitingTime = 2000
                    if (page != null) {
                        if (hash != page.hash) {
                            PageCache.instance.storeEntry(location, page)
                            continuousCompletion?.get()?.didUpdatePage(page)
                        }
                        waitingTime = 100
                    }
                    waiting = true
                    GlobalScope.launch(Dispatchers.Main, CoroutineStart.DEFAULT) {
                        delay(waitingTime.toLong())
                        waiting = false
                        tryNextContinuousLoad()
                    }
                }
            }
        }
    }

}

File: pirates-ios/PiratesGame/Page/Storage/PageLoader.swift
//
//  PageLoader.swift
//  Page storage: handles loading of pages from several sources
//

import Foundation
import Alamofire

protocol PageLoaderContinuousCompletion: class {
    
    func didUpdatePage(page: Page)
    
}

class PageLoader {
    
    // --
    // MARK: Members
    // --

    private let location: String
    private let loadInternal: Bool
    private weak var continuousCompletion: PageLoaderContinuousCompletion?
    private var loading = false
    private var waiting = false
    

    // --
    // MARK: Initialization
    // --
    
    init(location: String) {
        self.location = location
        loadInternal = !location.contains("://")
    }
    

    // --
    // MARK: Loading
    // --
    
    func load(completion: @escaping (_ page: Page?, _ error: Error?) -> Void) {
        if let page = PageCache.shared.getEntry(cacheKey: location) {
            completion(page, nil)
        } else if loadInternal {
            loadInternal(completion: { page, error in
                if let page = page {
                    PageCache.shared.storeEntry(cacheKey: self.location, page: page)
                }
                completion(page, error)
            })
        } else {
            loadOnline(currentHash: "ignore", completion: { page, error in
                if let page = page {
                    PageCache.shared.storeEntry(cacheKey: self.location, page: page)
                }
                completion(page, error)
            })
        }
    }
    
    private func loadOnline(currentHash: String, completion: @escaping (_ page: Page?, _ error: Error?) -> Void) {
        if !loadInternal {
            let headers: HTTPHeaders = [
                "X-Mock-Wait-Change-Hash": currentHash
            ]
            loading = true
            Alamofire.request(location, headers: headers).responseJSON { response in
                self.loading = false
                if let dictionary = response.value as? [String: Any] {
                    completion(Page(dictionary: dictionary, hash: response.data?.md5() ?? "unknown"), nil)
                } else {
                    completion(nil, response.error)
                }
            }
        } else {
            completion(nil, NSError(domain: "Loading mismatch", code: -1))
        }
    }
    
    private func loadInternal(completion: @escaping (_ page: Page?, _ error: Error?) -> Void) {
        loading = true
        DispatchQueue.global().async {
            let page = self.loadInternalSync()
            DispatchQueue.main.async {
                self.loading = false
                completion(page, page == nil ? NSError(domain: "Could not load page", code: -1) : nil)
            }
        }
    }
    
    func loadInternalSync() -> Page? {
        if loadInternal {
            if let path = Bundle.main.path(forResource: "Pages/" + location, ofType: "json") {
                if let jsonData = try? NSData(contentsOfFile: path, options: .mappedIfSafe) as Data {
                    return Page(jsonData: jsonData)
                }
            }
        }
        return nil
    }

   
    // --
    // MARK: Continuous loading
    // --
    
    func startLoadingContinuously(completion: PageLoaderContinuousCompletion) {
        continuousCompletion = completion
        tryNextContinuousLoad()
    }
    
    func stopLoadingContinuously() {
        continuousCompletion = nil
    }
    
    private func tryNextContinuousLoad() {
        if !loading && !waiting && continuousCompletion != nil {
            if loadInternal {
                loadInternal(completion: { page, error in
                    if let page = page {
                        PageCache.shared.storeEntry(cacheKey: self.location, page: page)
                        self.continuousCompletion?.didUpdatePage(page: page)
                    }
                })
            } else {
                let hash = PageCache.shared.getEntry(cacheKey: location)?.hash ?? "unknown"
                loadOnline(currentHash: hash, completion: { page, error in
                    var waitingTime: Double = 2
                    if let page = page {
                        if hash != page.hash {
                            PageCache.shared.storeEntry(cacheKey: self.location, page: page)
                            self.continuousCompletion?.didUpdatePage(page: page)
                        }
                        waitingTime = 0.1
                    }
                    self.waiting = true
                    DispatchQueue.main.asyncAfter(deadline: .now() + waitingTime, execute: {
                        self.waiting = false
                        self.tryNextContinuousLoad()
                    })
                })
            }
        }
    }

}
